OCPSTRAT-784: Simplify and unify adding nodes on day 2 using agent-based installer approach (#1517)
openshift/enhancements: enhancements/oc/day2-add-nodes.md

---
title: simplify-and-unify-day2-add-node
authors:
  - "@andfasano"
  - "@zaneb"

> approvers: 

Reviewer: From the [template](https://github.com/openshift/enhancements/blob/master/guidelines/enhancement_template.md?plain=1#L7):

> # A single approver is preferred, the role of the approver is to raise important questions, help ensure the enhancement receives reviews from all applicable areas/SMEs, and determine when consensus is achieved such that the EP can move forward to implementation.  Having multiple approvers makes it difficult to determine who is responsible for the actual approval.

  - "@patrickdillon"
  - "@oourfali"
  - "@soltysh"
reviewers: 
  - "@rwsu"
  - "@pawanpinjarkar"
creation-date: 2023-11-10
last-updated: 2023-11-10
tracking-link: 
  - "https://issues.redhat.com/browse/OCPSTRAT-784"
  - "https://issues.redhat.com/browse/AGENT-682"
see-also:
  - "https://github.com/openshift/enhancements/blob/master/enhancements/agent-installer/agent-based-installer.md"
replaces: []
superseded-by: []
---

# Simplify and unify adding nodes to clusters on day 2 using agent-based installer approach

## Summary
This enhancement proposal describes how to easily add nodes to any existing non-IPI cluster using lessons and technology from the agent-based 
installer. The user generates an ISO image via an `oc adm` command with an optional configuration file to describe the new nodes. The image 
is used to boot the hosts so that they will be added to the cluster without any further action, except for the final approval that remains a 
manual operation up to the user. An additional `oc adm` command will allow the user to monitor the progress of the task.

## Motivation

> The task of adding a node to an already existing non-IPI cluster (regardless of the installation approach adopted) still remains a complex 

Reviewer 2: If a cluster has been deployed via IPI but the user wants to use this method that should still be possible. For example you install with IPI on bare metal and then you add nodes but don't have access to their BMCs, which is a requirement for IPI bare metal deployment, this method solves this problem perfectly. So I wouldn't say this is for non-IPI clusters.

Author: The mentioned case is reported in the goals section, given that such specific scenario will be supported for IPI clusters, but in the other cases (especially for the public clouds) the expectation is that such approach wouldn't want to replace existing mechanisms (and also this point was added in the non-goals section)

Reviewer 2: Right, I think the non-IPI cluster mention makes it too restrictive and makes the reader think that adding nodes to an IPI-deployed cluster is out of the scope, when it shouldn't.

Author: Based also on @zaneb [comment](https://github.com/openshift/enhancements/pull/1517/files#r1413310649) (below), we could rephrase it so that it would be more clear the addressed scenarios/motivations. In general I'd like to emphasize that this proposal does not intend to address/replaced already working mechanism, such as the mentioned autoscaling

into different environments. 

So the current enhancement proposes a standard and unified way for adding nodes to an on-premises cluster, by replicating the same approach 
successfully adopted for the installation process, so that the end user could benefit from a more consistent and simple workflow. In addition, 
the standardization will make obsolete the hand-rolled UPI method, the need to use MCE to manage local clusters, and (unsupported) field-developed 
automations that are expensive to maintain.

### User Stories
- As a cluster administrator, I want to add (or replace) one or more nodes to an existing cluster (potentially deployed in a disconnected environment) 
  using the same simplified agent-based workflow already adopted for the installation
- As a cluster administrator, I want to provide the initial static network configuration required for each different host that will be added to the cluster
- As a cluster administrator, I want to validate the nodes to be added before initiating the procedure and be informed of any validation failures
- As a cluster administrator, I want to be able to monitor the process of adding new nodes

> ### Goals

Reviewer 2: Is there a list of platforms where this can work or not work initially?

Author: Mentioned the expected ones at [line 62](https://github.com/openshift/enhancements/pull/1517/files#diff-69aa9ff9e2b01a83a9d639863363dc17b351e16792d74d580142c0764056b6d4R62)

> - Allow the user to easily add one or more worker nodes to an existing (disconnected) non-IPI cluster, with the minimum amount of configuration requested

Reviewer 2: I'd say to any cluster, regardless of the installation method. While the focus is on non-IPI clusters, IPI clusters aren't discarded from this workflow (as you point out later with the use case where the user can't use BMCs in bare metal clusters) and we expect users to be able to use it (and they will use it for different reasons, no BMCs is one). So I'd reword this point to be less excluding of IPI clusters.

Reviewer: Maybe we say clusters where MachineSet scaleout is not available/possible. That's a more precise definition of what we're targeting.

Author: That sounds reasonable to me (see previous comment also)

- Provide a simpler interface by reusing existing command line tools for cluster administration
- When adding multiple nodes, a failure in adding a node should not impact the other nodes currently involved in the process
- Cover at least Baremetal, vSphere, Nutanix and None platform (regardless of the installation method used)
- For IPI clusters, provide support for the scenario where the user can’t use the hosts BMCs

### Non-Goals
- Replace MachineSet-based management of hosts for IPI clusters
- Offer more or finer-grained configuration options for new nodes than those supported by the agent-based installer
- To allow the creation of the ISO image for adding the nodes into an environment not directly connected to the target cluster
- Provide an automation for final nodes approval

## Proposal
A key factor for simplifying the user experience consists in offering a simple enough - but sufficiently flexible -  user interface 
to accomplish the task of adding the required nodes to a cluster, and to monitor the progress of such activity. For this reason, 
the proposal consists into enhancing the `oc` cli with two new commands:

> - `oc adm add-nodes-image` to generate the ISO for adding the nodes to the cluster

Reviewer: > IIUC, the generated image will be similar to the current agent.iso, but will include the functionality to invoke the '/import' API?

Correct.

> If so, it means that it won't work for the appliance use-case.

It won't at first, but not for that reason - it's no problem to put the actual code that invokes the /import API in the unconfigured ignition as well and have it conditionally triggered.

However, we'll have to have either e.g. an `add-nodes-config-image` command or e.g. a `--config-image` option to the `add-nodes-image` command (I prefer the latter) to support the appliance. This is probably quite doable, but we have to get the main thing working first.

> - `oc adm add-nodes-image` to generate the ISO for adding the nodes to the cluster

Reviewer 3: what about the timeline? will this functionality that will support the appliance be ready for 4.16?

> - `oc adm add-nodes-image` to generate the ISO for adding the nodes to the cluster

Reviewer 3: and what is the reason not to go for a solution that can work with the appliance from the start?

> - `oc adm add-nodes-image` to generate the ISO for adding the nodes to the cluster

Reviewer 4: We must have this functionality as part of 4.16. Let's make sure the approved enhancement includes everything required for it to work with the appliance, and have it as part of 4.16.

```

Where:
-`-o=’agent-nodes.iso`. The name of the ISO image to be generated.
-`–kubeconfig=’’`. Path to the kubeconfig file (or `KUBECONFIG` if available)

`nodes-config.yaml` is a configuration file, similar to `agent-config.yaml`, used to specify the list of the hosts to be added as worker 
nodes. For each host, it will be possible to specify a number of attributes, including a piece of nmstate configuration for the initial 
static network configuration. If not specified, the tool will look for a `nodes-config.yaml` file in the current directory.

The command will use the configured kubeconfig file, and the connection to the target cluster, to retrieve all the additional information 
required to validate and generate the ISO image that will be used for adding the configured nodes. 

#### nodes-config.yaml sample file
Following an example of `nodes-config.yaml` file, where a couple of workers are described with an initial static network configuration:

```
hosts:
- hostname: extra-worker-1
  rootDeviceHints:
   deviceName: /dev/sda
  interfaces:
   - macAddress: 00:00:00:00:00:00
     name: eth0
   - macAddress: 00:00:00:00:00:01

>      name: eth0

Reviewer 5: Duplicate `eth0` shown

- `/api/assisted-install/infra-envs/${INFRA_ENV_ID}/hosts/${HOST_ID}` to add the specified host to the cluster

#### oc integration notes
Since most of the existing code for agent-based installer approach is currently stored into the openshift/installer repo, the proposal is to implement the 
commands in a way similar to the already existing `oc adm must-gather` command, where essentially a pod is run within the target cluster to execute specific 
code build from the installer and shipped into a release image. Such approach will offer the following benefits:

- No code duplication/pollution. The required code for the additional commands will exist only in the openshift/installer repo (it will not be vendored into 
  the openshift/oc repo)
- Simplified integration. `oc` will take care only of marshaling the inputs/outputs of the command

To achieve this kind of loosely coupled level of integration with `oc`, a specific binary with the described commands will be built from the openshift/installer 
repo code and distributed into an image of the release payload as an additional artifact.

### Risks and Mitigations

> Most of the existing code for agent-based installer approach is currently stored into the openshift/installer repo (with additional vendoring of the api/client 

Reviewer 6: We are struggling with the maintenance issues(security scanners, code checkers, bugs, etc.) stemming from the vendored packages in oc and I think, we should be very reluctant to add more(installer, assisted-service). Since we already vendor openshift/api, this package can work.

Reviewer: This text is in the wrong section I think. It should be in the Alternatives section because vendoring all of the installer code into `oc` is something that we are explicitly not proposing to do.

`oc`'s job here (as described in the "oc integration notes" section above) is going to be to start a pod (a bit like the `must-gather` command) and stream out the ISO (again like `must-gather`) and the installer progress (like `oc logs`). It's syntactic sugar for stuff that `oc` is can already do.

@andfasano can you move this into the 'Alternatives' section to make it clear that we are _not_ planning to vendor a bunch of stuff into oc?

Reviewer 6: Thanks for the reply. Could you please add that which packages would be possibly needed to vendor in oc to see what are their dependency chain depths?.

Author: Agree with @zaneb , the section is not the proper one, will move it

Reviewer 6: Thanks for the clarification https://github.com/openshift/enhancements/pull/1517/files#diff-69aa9ff9e2b01a83a9d639863363dc17b351e16792d74d580142c0764056b6d4R205-R206

Reviewer: > Could you please add that which packages would be possibly needed to vendor in oc to see what are their dependency chain depths?.

I don't anticipate there being any at all.
It'll be purely doing stuff that oc can already do, like creating a Pod and streaming the output to a file/the console.

parts of openshift/assisted-service), and it wasn’t explicitly designed for being re-vendored. Vendoring the required parts in the openshift/oc repo will require

> a significant amount of refactoring in the openshift/install repo. Also, `oc` is not tied to a particular version, while the agent-based installer code is. 

Reviewer 6: oc has supported version skew +1/-1 guarantee and it seems that this may break that guarantee.

or not the required cluster infrastructure. The approach addressed several pain points, in particular by not requiring any additional provisioning 
hardware, and by applying a number of live pre-flight validations in the target environment to spot in advance those issues that could hamper 
the setup. Also, capturing all the necessary elements for the installation in a single ISO image simplified the application and distribution 
into different environments. 

The current enhancement proposes a standard and unified way for adding nodes to an on-premises cluster, by replicating the same approach 
successfully adopted for the installation process, so that the end user could benefit from a more consistent and simple workflow. In addition, 
the standardization will make obsolete the hand-rolled UPI method, the need to use MCE to manage local clusters, and (unsupported) field-developed 
automations that are expensive to maintain.

### User Stories
- As a cluster administrator, I want to add (or replace) one or more nodes to an existing cluster (potentially deployed in a disconnected environment) 

>   using the same simplified agent-based workflow already adopted for the installation

Author: Ok

the proposal consists into enhancing the `oc` cli with two new commands:

- `oc adm add-nodes-image` to generate the ISO for adding the nodes to the cluster
- `oc adm monitor-add-nodes` to monitor the nodes currently being added to the cluster

There are several reasons for choosing `oc` as a primary user interfaces for these tasks:
- It’s the official cluster administration tool, so it is expected to be already available in the target cluster environment
- It already supports the kubeconfig file for interacting with the target cluster (it’s relevant since it contains several pieces of 
  information required for adding the nodes, as explained in the design section)
- It already offers a couple of commands to manually approve/deny certificate signing requests (CSRs) when adding new nodes

The usage of `oc` will also help in reducing the amount of information that the user will need to provide, resulting thus in a simplified
input configuration file.

### add-nodes-image command
The format of the command will be the following:

```
oc -–kubeconfig=’’ adm add-nodes-image [-o=’agent-nodes.iso’] <nodes-config.yaml> 
```

Where:

> -`-o=’agent-nodes.iso`. The name of the ISO image to be generated.

Reviewer: ```suggestion
- `-o='agent-nodes.iso'`. The name of the ISO image to be generated.
```

> -`–kubeconfig=’’`. Path to the kubeconfig file (or `KUBECONFIG` if available)

Reviewer: ```suggestion
- `--kubeconfig=''`. Path to the kubeconfig file (or `KUBECONFIG` if available)
```

> `nodes-config.yaml` is a configuration file, similar to `agent-config.yaml`, used to specify the list of the hosts to be added as worker 

Reviewer: I think it's important that this file be optional, because having to construct it is a big barrier to entry. The common case will be adding one or more nodes with no network configuration. That should be possible without passing any config.

I think it would also be good to have separate `--network-config=<nmstate.yaml>` and `--root-device=<hint>:<value>` arguments that apply to all nodes booted from the ISO.

And then the nodes-config.yaml would be used only for the case where you are simultaneously adding multiple nodes with static IPs.

Author: > I think it's important that this file be optional, because having to construct it is a big barrier to entry. The common case will be adding one or more nodes with no network configuration. That should be possible without passing any config.

These seems reasonable, especially I guess to cover the DHCP scenario, not sure anyhow if the hostnames/macs could be avoided at all. I agree it would be important to make easy the more common scenarios (ie add a node without a specific config)

> I think it would also be good to have separate --network-config=<nmstate.yaml> and --root-device=<hint>:<value> arguments that apply to all nodes booted from the ISO.

Having (some of ) the different pieces captured by the single config file `nodes-config.yaml` seems to me a more complicated interface for the end user, especially in the case of more than one node to be added - and in particular if a `nmstate.yaml` file has to be provided. 

> And then the nodes-config.yaml would be used only for the case where you are simultaneously adding multiple nodes with static IPs.

The main role of the `nodes-config.yaml` is to act as a public API for the end user, since it defines what could be configured or not for a node to be added. It could be also easily expanded in case of need, and stored in a vcs when required.

#### nodes-config.yaml sample file
Following an example of `nodes-config.yaml` file, where a couple of workers are described with an initial static network configuration:

```
hosts:
- hostname: extra-worker-1
  rootDeviceHints:
   deviceName: /dev/sda
  interfaces:
   - macAddress: 00:00:00:00:00:00
     name: eth0
   - macAddress: 00:00:00:00:00:01
     name: eth1
  networkConfig:
   interfaces:
     - name: eth0
       type: ethernet
       state: up
       mac-address: 00:00:00:00:00:00
       ipv4:
         enabled: true
         address:
           - ip: 192.168.122.2
             prefix-length: 23
         dhcp: false
- hostname: extra-worker-2
  rootDeviceHints:
   deviceName: /dev/sda
  interfaces:
   - macAddress: 00:00:00:00:00:02
     name: eth0
  networkConfig:
   interfaces:
     - name: eth0
       type: ethernet
       state: up
       mac-address: 00:00:00:00:00:02
       ipv4:
         enabled: true
         address:
           - ip: 192.168.122.3
             prefix-length: 23
         dhcp: false
```

### monitor-add-nodes command

The format of the command will be the following:

> `oc -–kubeconfig=’’ adm monitor-add-nodes <IPs list>`

Reviewer: ```suggestion
`oc --kubeconfig='' adm monitor-add-nodes <IPs list>`
```

Where:

> - `–kubeconfig=’’``. Path to the kubeconfig file (or KUBECONFIG if available)

Reviewer: ```suggestion
- `--kubeconfig=''`. Path to the kubeconfig file (or KUBECONFIG if available)
```

> - `IPs list`` is a list of the nodes IP to be monitored

Reviewer: ```suggestion
- `<IPs list>` is a list of the nodes IP to be monitored
```

### Workflow Description
The working environment must satisfy the initial following conditions:
- `oc` tool is available
- A direct connection to the target cluster is available
- kubeconfig file is available

The following steps illustrates the expected workflow:
1. The cluster administrator prepares the `nodes-config.yaml` file listing all the nodes to be added, optionally with their initial network configuration
2. The cluster administrator runs the `oc adm add-nodes-image` command to generate the ISO
3. The cluster administrator boots every node previously listed using the freshly generated ISO
4. The cluster administrator runs the `oc adm monitor-add-nodes` command to monitor the activity progress
5. Once a node csr is pending, the cluster administrator runs the `oc adm certificate approve <csr-id>` to complete the process for the specified node

### API Extensions
N/A

### Implementation Details/Notes/Constraints
The implementation will largely reuse the same approach / code present in the openshift/installer repo for creating an ISO image 
for the installation, with a number of exceptions that will be detailed in the subsequent paragraphs.

#### Proposed architecture

> Every node booted with the generated ISO will be independent for adding itself to the cluster - without any need to specify a single rendezvous IP, 

Reviewer: suggest removing the comma at the end here, as it makes it slightly ambiguous in my reading.

Author: Ack

as it happens for the installation scenario. In this way if a node is failing for some reason (for example, the pre-flight validations block the process), 
the other nodes may continue the joining process undisturbed.
The following hilights better the proposed architecture:

![alt Architecture diagram](day2-add-nodes-diagram.png)

#### Assisted-service interaction
The following API calls will be used to add a single node to the target cluster:
- `/api/assisted-install/v2/clusters/import` to import an existing cluster
- `/api/assisted-install/infra-envs/${INFRA_ENV_ID}/hosts/${HOST_ID}` to add the specified host to the cluster

> #### oc integration notes

Reviewer: I wonder if there's a way to structure things so that this section appears before the detailed description of the `oc` command syntax.

Author: I put this part here since it was under the `Implementation Details/Notes/Constraints` section. Maybe we could add a note in the `oc` commands description that the integration details will be provided later in the proposal

### User Stories
- As a cluster administrator, I want to add (or replace) one or more nodes to an existing cluster (potentially deployed in a disconnected environment) 
  by generating an ISO and use it to boot the required nodes
- As a cluster administrator, I want to provide the initial static network configuration required for each different host that will be added to the cluster
- As a cluster administrator, I want to validate the nodes to be added before initiating the procedure and be informed of any validation failures
- As a cluster administrator, I want to be able to monitor the process of adding new nodes

### Goals
- Allow the user to easily add one or more worker nodes to an existing (disconnected) cluster where MachineSet scaleout is not available or feasible, 
  with the minimum amount of configuration requested
- Provide a simpler interface by reusing existing command line tools for cluster administration
- When adding multiple nodes, a failure in adding a node should not impact the other nodes currently involved in the process
- Cover at least Baremetal, vSphere, Nutanix and None platform (regardless of the installation method used)

> - For IPI clusters, provide support for the scenario where the user can’t use the hosts BMCs

Reviewer 7: To clarify because it was in its own bullet point, to support IPI clusters where users can't use the hosts' BMCs, the workflow isn't any different compared to the other platforms being covered.

### Non-Goals
- Replace MachineSet-based management of hosts for IPI clusters
- Offer more or finer-grained configuration options for new nodes than those supported by the agent-based installer
- Generation of the ISO image for adding the nodes into an environment not directly connected to the target cluster
- Provide an automation for final nodes approval

## Proposal
A key factor for simplifying the user experience consists in offering a simple enough - but sufficiently flexible -  user interface 
to accomplish the task of adding the required nodes to a cluster, and to monitor the progress of such activity. For this reason, 
the proposal consists into enhancing the `oc` cli with two new commands:

- `oc adm add-nodes-image` to generate the ISO for adding the nodes to the cluster

> - `oc adm monitor-add-nodes` to monitor the nodes currently being added to the cluster

Reviewer 8: I'd consider introducing a single command with either separate flags or sub-commands handling adding vs monitoring. Maybe something like `oc adm add-node-images` which would either check for an existing namespace and report progress or create a new invocation. I'm suspecting that don't want to allow running more than a single instance at any point in time. Additionally some kind of `--wait` flag which would allow the user to just stream/watch the logs as it goes.

Reviewer: That's definitely worth considering, but the reason for splitting it up is that there is a manual step (booting the ISO you got in the first part) in between the two parts.
Furthermore, there is data required as input to the second part that may not be available when the first part is invoked - if you are using DHCP then the easiest time to get the IP address is _after_ you've booted the server with the ISO.

I don't think running multiple instances would hurt, the more important thing is that we clean them up when they're done.

(More specific details about the proposed integration with `oc` could be find later in the [integration notes](#oc-integration-notes) paragraph)

There are several reasons for choosing `oc` as a primary user interfaces for these tasks:
- It’s the official cluster administration tool, so it is expected to be already available in the target cluster environment

> - It already supports the kubeconfig file for interacting with the target cluster (it’s relevant since it contains several pieces of 

Reviewer: `openshift-installer` doesn't, it uses an install-config. Putting this functionality in the installer was a serious competing proposal that we rejected for this reason.

input configuration file.

### add-nodes-image command
The format of the command will be the following:

```
oc --kubeconfig='' adm add-nodes-image [-o='agent-nodes.iso'] <nodes-config.yaml> 
```

Where:
- `-o='agent-nodes.iso'`. The name of the ISO image to be generated.
- `--kubeconfig=''`. Path to the kubeconfig file (or `KUBECONFIG` if available)

`nodes-config.yaml` is a configuration file, similar to `agent-config.yaml`, used to specify the list of the hosts to be added as worker 
nodes. For each host, it will be possible to specify a number of attributes, including a piece of nmstate configuration for the initial 
static network configuration. If not specified, the tool will look for a `nodes-config.yaml` file in the current directory.

The command will use the configured kubeconfig file, and the connection to the target cluster, to retrieve all the additional information 
required to validate and generate the ISO image that will be used for adding the configured nodes. 

#### nodes-config.yaml sample file
Following an example of `nodes-config.yaml` file, where a couple of workers are described with an initial static network configuration:

```

> hosts:

Reviewer 9: Other inputs, including the install-config, look more like a kube API resource with version, group, kind, and other metadata. It would make sense to take that approach here, too, in case the schema of the input file needs to change in the future.

Is this structure definition going to be more or less the same as the one used in the installer for the same data? Can we share that code?

Reviewer: > It would make sense to take that approach here, too, in case the schema of the input file needs to change in the future.

There's definitely a usability tradeoff here, in making the user include a bunch of boilerplate.
I'm not actually convinced the versioning does anything in this case. The thing that is effective is being disciplined about how you change fields. It's possible that using the k8s-style metadata acts as a reminder to developers about what can and cannot be safely changed, but it seems to me that's the main effect.

Versioning allows you to change the type of a field without changing the name of it. But that can be worked around by creating a new field with a different name - which you should probably do anyway. It also allows you to remove a field while still supporting old config files. But since you're still supporting the old field anyway there's no point creating a new version where the field is forbidden.

In practice the install-config version never changes and we just keep adding and deprecating fields without removing anything.

- hostname: extra-worker-1
  rootDeviceHints:
   deviceName: /dev/sda
  interfaces:
   - macAddress: 00:00:00:00:00:00
     name: eth0
   - macAddress: 00:00:00:00:00:01
     name: eth1
  networkConfig:
   interfaces:
     - name: eth0
       type: ethernet
       state: up
       mac-address: 00:00:00:00:00:00
       ipv4:
         enabled: true
         address:
           - ip: 192.168.122.2
             prefix-length: 23
         dhcp: false
- hostname: extra-worker-2
  rootDeviceHints:
   deviceName: /dev/sda
  interfaces:
   - macAddress: 00:00:00:00:00:02
     name: eth0
  networkConfig:
   interfaces:
     - name: eth0
       type: ethernet
       state: up
       mac-address: 00:00:00:00:00:02
       ipv4:
         enabled: true
         address:
           - ip: 192.168.122.3
             prefix-length: 23
         dhcp: false
```

### monitor-add-nodes command

The format of the command will be the following:

> `oc --kubeconfig='' adm monitor-add-nodes <IPs list>`

Reviewer 7: One complication I could see here is if users used dhcp in their nmstate configs and if they have no way of knowing the IPs that will be assigned.

Reviewer: They don't have to run this command until after the host is booted, so they should be able to know everything they are going to know at that point.

Where:
- `--kubeconfig=''`. Path to the kubeconfig file (or KUBECONFIG if available)
- `<IPs list>` is a list of the nodes IP to be monitored

> The command will try to connect to each specified node to retrieve initially the results of the pre-flight validations, and then 

Reviewer: It's an HTTP API, but the connection will come from the same binary that generates the ISO (i.e. running inside the cluster).

  testing SNO/compact/HA topologies. These workflows are currently used by many different openshift/release repos and by release controller agent informing jobs.
  The workflows will be extended with a config var to allow exercising the task of adding a new node, using the method described.

### Graduation Criteria

#### Dev Preview -> Tech Preview
N/A

#### Tech Preview -> GA
N/A

#### Removing a deprecated feature
N/A

### Upgrade / Downgrade Strategy
N/A

### Version Skew Strategy
N/A

### Operational Aspects of API Extensions
N/A

#### Failure Modes
N/A

#### Support Procedures
N/A

## Implementation History
- N/A

## Alternatives
- It would be possible to create an automated UPI image that automatically starts the installation via chaining the Ignition generated by MCO in the cluster, 
  without using at all assisted-service. This would mean not offering to the end user the pre-flight validations, which are a key element for helping the user in 
  real case environments where not all the elements could be known upfront
- Using `oc` plugin mechanism to implement the required commands. From a UX point of view this approach anyhow does not seem the suitable one due the various 

>   limitations present: a plugin cannot be integrated as a sub-command (ie, to `oc adm`) - resulting then in a non-uniform administrative interface, and the

Reviewer 6: I think we can customize oc additionally enabling this for `oc adm`. But that brings about separate installation of the tool in addition to oc. I think, it is not desirable in this KEP.
